File: environment/src/main/java/jetbrains/exodus/env/management/EnvironmentStatistics.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.env.management;

import jetbrains.exodus.core.dataStructures.ObjectCacheBase;
import jetbrains.exodus.env.Environment;
import jetbrains.exodus.env.EnvironmentImpl;
import jetbrains.exodus.env.EnvironmentStatistics.Type;
import jetbrains.exodus.management.MBeanBase;
import jetbrains.exodus.management.Statistics;
import org.jetbrains.annotations.NotNull;

import static jetbrains.exodus.env.EnvironmentStatistics.Type.*;

public class EnvironmentStatistics extends MBeanBase implements EnvironmentStatisticsMBean {

    @NotNull
    private final EnvironmentImpl env;
    @NotNull
    private final Statistics<Type> statistics;

    @SuppressWarnings("unchecked")
    public EnvironmentStatistics(@NotNull final EnvironmentImpl env) {
        super(getObjectName(env));
        this.env = env;
        statistics = env.getStatistics();
    }

    static String getObjectName(@NotNull final Environment env) {
        return OBJECT_NAME_PREFIX + ", location=" + escapeLocation(env.getLocation());
    }

    @Override
    public long getBytesWritten() {
        return getTotal(BYTES_WRITTEN);
    }

    @Override
    public double getBytesWrittenPerSecond() {
        return getMean(BYTES_WRITTEN);
    }

    @Override
    public long getBytesRead() {
        return getTotal(BYTES_READ);
    }

    @Override
    public double getBytesReadPerSecond() {
        return getMean(BYTES_READ);
    }

    @Override
    public long getBytesMovedByGC() {
        return getTotal(BYTES_MOVED_BY_GC);
    }

    @Override
    public double getBytesMovedByGCPerSecond() {
        return getMean(BYTES_MOVED_BY_GC);
    }

    @Override
    public String getLogCacheHitRate() {
        return ObjectCacheBase.formatHitRate((float) getMean(LOG_CACHE_HIT_RATE));
    }

    @Override
    public long getNumberOfTransactions() {
        return getTotal(TRANSACTIONS);
    }

    @Override
    public double getNumberOfTransactionsPerSecond() {
        return getMean(TRANSACTIONS);
    }

    @Override
    public long getNumberOfReadonlyTransactions() {
        return getTotal(READONLY_TRANSACTIONS);
    }

    @Override
    public double getNumberOfReadonlyTransactionsPerSecond() {
        return getMean(READONLY_TRANSACTIONS);
    }

    @Override
    public int getActiveTransactions() {
        return (int) getTotal(ACTIVE_TRANSACTIONS);
    }

    @Override
    public long getNumberOfFlushedTransactions() {
        return getTotal(FLUSHED_TRANSACTIONS);
    }

    @Override
    public double getNumberOfFlushedTransactionsPerSecond() {
        return getMean(FLUSHED_TRANSACTIONS);
    }

    @Override
    public long getDiskUsage() {
        return getTotal(DISK_USAGE);
    }

    @Override
    public int getUtilizationPercent() {
        return (int) getTotal(UTILIZATION_PERCENT);
    }

    @Override
    public String getStoreGetCacheHitRate() {
        return ObjectCacheBase.formatHitRate((float) getMean(STORE_GET_CACHE_HIT_RATE));
    }

    @Override
    public String getStuckTransactionMonitorMessage() {
        return env.getStuckTransactionMonitorMessage();
    }

    private long getTotal(@NotNull final Type statisticsName) {
        return statistics.getStatisticsItem(statisticsName).getTotal();
    }

    private double getMean(@NotNull final Type statisticsName) {
        return statistics.getStatisticsItem(statisticsName).getMean();
    }
}

File: environment/src/main/kotlin/jetbrains/exodus/env/management/EnvironmentStatistics.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.env.management

import jetbrains.exodus.env.Environment
import jetbrains.exodus.env.EnvironmentImpl
import jetbrains.exodus.env.EnvironmentStatistics
import jetbrains.exodus.management.MBeanBase

class EnvironmentStatistics(private val env: EnvironmentImpl) : MBeanBase(getObjectName(env)),
    EnvironmentStatisticsMBean {

    private fun getStatistics() = env.statistics

    override fun getBytesWritten(): Long = getTotal(EnvironmentStatistics.Type.BYTES_WRITTEN)

    override fun getBytesWrittenPerSecond(): Double = getMean(EnvironmentStatistics.Type.BYTES_WRITTEN)

    override fun getBytesRead(): Long = getTotal(EnvironmentStatistics.Type.BYTES_READ)

    override fun getBytesReadPerSecond(): Double = getMean(EnvironmentStatistics.Type.BYTES_READ)

    override fun getBytesMovedByGC(): Long = getTotal(EnvironmentStatistics.Type.BYTES_MOVED_BY_GC)

    override fun getBytesMovedByGCPerSecond(): Double = getMean(EnvironmentStatistics.Type.BYTES_MOVED_BY_GC)

    override fun getLogCacheHitRate(): Float = env.log.getCacheHitRate()

    override fun getNumberOfTransactions(): Long = getTotal(EnvironmentStatistics.Type.TRANSACTIONS)

    override fun getNumberOfTransactionsPerSecond(): Double = getMean(EnvironmentStatistics.Type.TRANSACTIONS)

    override fun getNumberOfReadonlyTransactions(): Long = getTotal(EnvironmentStatistics.Type.READONLY_TRANSACTIONS)

    override fun getNumberOfReadonlyTransactionsPerSecond(): Double =
        getMean(EnvironmentStatistics.Type.READONLY_TRANSACTIONS)

    override fun getNumberOfGCTransactions(): Long = getTotal(EnvironmentStatistics.Type.GC_TRANSACTIONS)

    override fun getNumberOfGCTransactionsPerSecond(): Double = getMean(EnvironmentStatistics.Type.GC_TRANSACTIONS)

    override fun getActiveTransactions(): Int = getTotal(EnvironmentStatistics.Type.ACTIVE_TRANSACTIONS).toInt()

    override fun getNumberOfFlushedTransactions(): Long = getTotal(EnvironmentStatistics.Type.FLUSHED_TRANSACTIONS)

    override fun getNumberOfFlushedTransactionsPerSecond(): Double =
        getMean(EnvironmentStatistics.Type.FLUSHED_TRANSACTIONS)

    override fun getTransactionsDuration(): Long = getTotal(EnvironmentStatistics.Type.TRANSACTIONS_DURATION)

    override fun getReadonlyTransactionsDuration(): Long =
        getTotal(EnvironmentStatistics.Type.READONLY_TRANSACTIONS_DURATION)

    override fun getGcTransactionsDuration(): Long = getTotal(EnvironmentStatistics.Type.GC_TRANSACTIONS_DURATION)

    override fun getDiskUsage(): Long = getTotal(EnvironmentStatistics.Type.DISK_USAGE)

    override fun getUtilizationPercent(): Int = getTotal(EnvironmentStatistics.Type.UTILIZATION_PERCENT).toInt()

    override fun getStoreGetCacheHitRate(): Float = env.getStoreGetCacheHitRate()

    override fun getStuckTransactionCount(): Int = env.getStuckTransactionCount()

    private fun getTotal(statisticsName: EnvironmentStatistics.Type): Long {
        return getStatistics().getStatisticsItem(statisticsName).total
    }

    private fun getMean(statisticsName: EnvironmentStatistics.Type): Double {
        return getStatistics().getStatisticsItem(statisticsName).mean
    }

    companion object {
        internal fun getObjectName(env: Environment) =
            "$STATISTICS_OBJECT_NAME_PREFIX, location=${escapeLocation(env.location)}"
    }

}
